core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala

```diff
@@ -344,12 +354,20 @@ private[spark] class MesosSchedulerBackend(
 
   override def frameworkMessage(d: SchedulerDriver, e: ExecutorID, s: SlaveID, b: Array[Byte]) {}
 
+  /**
+   * Remove executor associated with slaveId in a thread safe manner.
+   */
+  private def removeExecutor(slaveId: String) = {
+    synchronized {
```

Comment:
Can you explain a bit what this synchronization is trying to protect? I did a small audit of this class and I found the synchronization is a bit all over the place. For instance we don't seem to hold any locks when adding to `slaveIdsWithExecutors` but we do hold a lock when removing from it. Seems a bit odd.